archive/src/expand/all.rs

```diff
@@ -0,0 +1,204 @@
+    /// upon which the operation was attempted.
+    #[getset(get = "pub")]
+    warnings: HashMap<Source, Vec<Error>>,
+}
+
+impl Drop for Expansion {
+    fn drop(&mut self) {
+        let _ = self.cleanup();
+    }
+}
+
+impl Expansion {
+    /// Persist the temporary directories to disk, returning the contents of `Self` as a tuple.
+    ///
+    /// This consumes the `Expansion` without deleting temporary directories on the file system,
+    /// meaning that they are no longer automatically deleted.
+    pub fn persist(self) -> (BiHashMap<Source, Destination>, HashMap<Source, Vec<Error>>) {
+        let mut this = mem::ManuallyDrop::new(self);
+        (
+            mem::take(&mut this.locations),
+            mem::take(&mut this.warnings),
+        )
+    }
+
+    /// Delete all destinations and clear the `locations` map.
+    ///
+    /// If no errors are encountered the result is ok;
+    /// if any errors are encountered they are collected into the returned error list.
+    ///
+    /// This function is equivalent to causing `Expansion` to be dropped by letting it go out of scope;
+    /// the intention with this function is to enable checking for cleanup errors when desired.
+    ///
+    /// It is supported to call `cleanup` multiple times; subsequent operations are a no-op,
+    /// regardless whether the first call to `cleanup` was fully successful.
+    pub fn cleanup(&mut self) -> Result<(), Vec<Error>> {
+        // Special case for when dropped after manually running `cleanup`.
+        if self.locations.is_empty() {
+            return Ok(());
+        }
+
+        let errors = self
+            .locations
+            .right_values()
+            .map(|d| d.inner().to_owned())
+            .fold(Vec::new(), |mut acc, destination| {
+                if let Err(error) = fs::remove_dir_all(&destination) {
+                    acc.push(Error::Cleanup { destination, error });
+                }
+                acc
+            });
+
+        self.locations = BiHashMap::default();
+        if errors.is_empty() {
+            Ok(())
+        } else {
+            Err(errors)
+        }
+    }
+
+    /// List all destinations.
+    pub fn destinations(&self) -> HashSet<Destination> {
+        self.locations.right_values().cloned().collect()
+    }
+
+    fn record(&mut self, attempt: Attempt) {
+        match attempt.result {
+            Ok(destination) => {
+                self.locations
+                    .insert(attempt.source.into(), destination.into());
+            }
+            Err(err) => {
+                self.warn(attempt.source, err.into());
+            }
+        }
+    }
+
+    fn warn(&mut self, source: PathBuf, warning: Error) {
+        let errs = self.warnings.entry(Source::from(source)).or_insert(vec![]);
+        errs.push(warning);
+    }
+
+    fn record_many(&mut self, attempts: impl Iterator<Item = Attempt>) {
+        attempts.into_iter().for_each(|a| self.record(a))
+    }
```

Comment:
[nit]: There's a collision of `Iterator` and `IntoIterator` here.  Since `attempts` implements `Iterator` directly, the `into_iter` here is a no-op.  I think we actually want to receive `impl IntoIterator<_>` here:
```suggestion
    fn record_many(&mut self, attempts: impl IntoIterator<Item = Attempt>) {
        attempts.into_iter().for_each(|a| self.record(a))
    }
```

Which allows `expansion.record_many(vec![]);` to compile, without needing `into_iter` at the call site.